xpsparser/parser.py

```diff
@@ -22,42 +22,580 @@
+                else:
+                    setattr(metadata, 'spectrum_region', self.current_region)
+                if 'Group' in dictionary.keys():
+                    self.current_group = dictionary['Group']
+                else:
+                    setattr(metadata, 'group_name', self.current_group)
+        
+        """ Get what can be found in the global header."""
+        for key in self.global_header.keys():
+            if key in self.group_metadata_attribute_map.keys():
+                attr = self.group_metadata_attribute_map[key]
+                value = self.global_header[key]
+                setattr(metadata, attr, value)
+                
+        """ Then get from kwargs."""
+        for key in self.kwarg_keys:
+            if key in self.__dict__.keys():
+                setattr(metadata, key, getattr(self, key))
+        
+        """ Then get method type."""
+        method_type = self._getMethodType(group)
+        metadata.method_type = method_type
+                
+        return metadata
+
+    def _getDeviceSettings(self, channel, device_settings_object):
+        """ Get the device settings."""
+        
+        """ First look in the channel's data."""
+        settings = device_settings_object
+        for key in channel[0].keys():
+            if key in self.settings_attribute_map.keys():
+                attr = self.settings_attribute_map[key]
+                value = channel[0][key]
+                setattr(settings, attr, value)
+        
+        """ Then look in the global header."""
+        for key in self.global_header.keys():
+            if key in self.settings_attribute_map.keys():
+                attr = self.settings_attribute_map[key]
+                value = self.global_header[key]
+                setattr(settings, attr, value)    
+                
+        settings.device_name = self._getDeviceName(channel)
+        
+        return settings
+    
+    def _getDeviceName(self, channel):
+        """ Get the name of the device."""
+        
+        """ The device name is not provided for the spectrometer, so 
+        get it from the default device name.
+        """
+        if channel[0]['channel_type'] == 'primary':
+            device_name = self.default_primary_device_name
+        elif channel[0]['channel_type'] == 'axis':
+            device_name = self.default_axis_device_name
+        else:
+            for key in channel[0].keys():
+                if 'External Channel' in key:
+                    for _device_name in self.known_device_names.keys():
+                        if _device_name in channel[0][key]:
+                            device_name = self.known_device_names[_device_name]
+                        else:
+                            device_name = 'unknown device'
+        
+        return device_name
+    
+    def _addDataChannels(self, group, measurement_data):
+        """ Add all the data channels to the measurement data.
+        PARAMETERS:
```

Comment:
Mark's code does not adhere to our style conventions. We are using google style docstrings:
```
'''
This does something.

Arguments:
    - foo: Foo is denoting bar.

Returns:
    The result of the function.
```